petpals/messaging/routes.py

```diff
@@ -0,0 +1,193 @@
+        'messages.html',
+        recent_messages=recent_messages,
+        current_user=current_user,
+        active_page=active_page,
+    )
+
+
+@router.get('/new')
+def new_message():
+    usernames = []
+    next_conversation_id = 0
+    conversation_id_query = db.session.query(func.max(Messages.conversation_id)).all()
+
+    if conversation_id_query[0][0] != None:
+        next_conversation_id = (
+            int(db.session.query(func.max(Messages.conversation_id)).all()[0][0]) + 1
+        )
+    else:
+        next_conversation_id = 1
+
+    users = User.query.all()
+    messages = Messages.query.all()
+
+    for user in users:
+        if current_user.username != user.username:
+            usernames.append(user.username)
+
+    for message in messages:
+        if current_user.username == message.sender_username:
+            if message.recipient_username in usernames:
+                usernames.remove(message.recipient_username)
+        elif current_user.username == message.recipient_username:
+            if message.sender_username in usernames:
+                usernames.remove(message.sender_username)
+
+    return render_template(
+        'new_message.html',
+        usernames=sorted(usernames),
+        sender=current_user.username,
+        active_page=active_page,
+        conversation_id=next_conversation_id,
+    )
+
+
+@router.post('/new/<recipient>')
+def send_new_message(recipient):
+    data = request.json
+
+    sender_checker = (
+        Messages.query.with_entities(
+            Messages.sender_username,
+            Messages.recipient_username,
+            Messages.conversation_id,
+        )
+        .filter_by(sender_username=data['sender'], recipient_username=data['recipient'])
+        .all()
+    )
+    recipient_checker = (
+        Messages.query.with_entities(
+            Messages.sender_username,
+            Messages.recipient_username,
+            Messages.conversation_id,
+        )
+        .filter_by(sender_username=data['recipient'], recipient_username=data['sender'])
+        .all()
+    )
+
+    new_message = Messages()
+
+    if db.session.query(func.max(Messages.conversation_id)).all()[0][0] == None:
+        new_message.conversation_id = 1
+    elif len(sender_checker) > 0:
+        new_message.conversation_id = sender_checker[0][2]
+    elif len(recipient_checker) > 0:
+        new_message.conversation_id = recipient_checker[0][2]
+    else:
+        new_message.conversation_id = (
+            int(db.session.query(func.max(Messages.conversation_id)).all()[0][0]) + 1
+        )
+
+    new_message.recipient_username = data['recipient']
+    new_message.sender_username = data['sender']
+    new_message.message = data['message']
+    new_message.time_sent = datetime.now()
+
+    db.session.add(new_message)
+    db.session.commit()
+    return recipient
+
+
+@router.get("/conversation/<conversation_id>")
+def get_conversation(conversation_id):
+    recipient = ""
+
+    time.sleep(1)
```

Comment:
No.

Reply:
It looks bad, but I have to declare 'recipient' so that it's in the correct scope.  The 'sleep' function is there to avoid loading the page before the data is inserted into the DB, then queried to be placed on the webpage.

Reply:
Fetch isn't asynchronous, it should be fine. If it's not fine, using a sleep is a very hacky way that is not safe and doesn't guarantee anything.